keps/sig-node/3721-support-for-env-files/README.md

```diff
@@ -0,0 +1,546 @@
+Managing unique configs for each Pod in Kubernetes can be cumbersome. Traditionally,
+Kubernetes ConfigMap are used to store configs and populate env vars,
+but this approach becomes less efficient when dealing with per-pod configs.
+
+To simplify this process, Kubernetes can introduce a new mechanism where an `initContainer`
+can build a config file. The main application container then utilizes
+the `fileKeyRef` field to reference this file and populate its env vars. Kubernetes
+automatically parses the file and sets the env vars during container creation without
+mounting the file directly.
+
+This enhancement offers a more manageable way to handle per-pod config in Kubernetes.
+In the example below, the Pod's output includes `CONFIG_MAIN=hello`.
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: dapi-test-pod
+spec:
+  initContainers:
+  - name: setup-envfile
+    image: registry.k8s.io/busybox
+    command: ['sh', '-c', 'echo "CONFIG_INIT=hello" > /data/config.env']
+    volumeMounts:
+    - name: config
+      mountPath: /data
+   containers:
+  - name: use-envfile
+    image: registry.k8s.io/busybox
+    command: [ "/bin/sh", "-c", "env" ]
+    env:
+    - name: CONFIG_MAIN
+      valueFrom:
+        fileKeyRef:
+          path: config.env
+          volumeName: config
+          key: CONFIG_INIT
+  restartPolicy: Never
+  volumes:
+  - name: config
+    emptyDir: {}
+```
+
+
+
+## Design Details
+
+Pods and PodTemplate will include `FileKeySelector` field under [`EnvVarSource`](https://pkg.go.dev/k8s.io/api/core/v1#EnvVarSource)
+that can be set for an individual container.
+
+### Pod API
+
+`FileKeySelector` field will allow the users to select the value for environment
+variable from the specified key in the file. The user can specify a file using `VolumeName` and
+`Path` fields. The `VolumeName` field specifies the volume mount that contains
+the file and `Path` is the relative path in this volume mount filesystem.
+
+
+```
+type EnvVarSource struct {
+    ...
+    // Selects a key of the env file.
+    // +optional
+    FileKeyRef *FileKeySelector `json:"fileKeyRef,omitempty" protobuf:"bytes,5,opt,name=fileKeyRef"`
+    ...
+}
+
+type FileKeySelector struct {
+    // The name of the volume mount containing the env file.
+    VolumeName string `json:",inline" protobuf:"bytes,1,opt,name=volumeName"`
+    // The relative file path inside the volume mount to select from.
+    Path string `json:",inline" protobuf:"bytes,2,opt,name=path"`
+    // The key of the env file to select from. It must be a valid key within the
+    // env file. An invalid key will prevent the pod from starting, even if it's
+    // successfully created.
+    Key string `json:"key" protobuf:"bytes,3,opt,name=key"`
+    // Specify whether the file or its key must be defined. If the file or key
+    // does not exist, then the env var is not published.
+    // +optional
+    Optional *bool `json:"optional,omitempty" protobuf:"varint,4,opt,name=optional"`
+}
+```
+
```

Comment:
```suggestion
Note: Container can mount the volume with the file and update it after started. This will affect the next start of a container if it has crashed.
```